### PR TITLE
util/kvencoder: use reference count to keep single domain instance

### PR DESCRIPTION
**Pull request by tiancaiamao**
## What have you changed? (mandatory)

BootstrapSession should not be called many times and we should keep
just a single domain instance

## What is the type of the changes? (mandatory)

- Improvement (non-breaking change which is an improvement to an existing feature)

## How has this PR been tested? (mandatory)

unit test

@winkyao

### CODE REVIEW

File: util/kvencoder/kv_encoder.go [hunk 1]
```diff
@@ -77,27 +77,43 @@ type KvEncoder interface {
...
+	se    session.Session
 	store kv.Storage
 	dom   *domain.Domain
-	se    session.Session
 }
 
 // New new a KvEncoder
 func New(dbName string, idAlloc autoid.Allocator) (KvEncoder, error) {
 	kvEnc := &kvEncoder{}
+	if atomic.LoadInt64(&refCount) == 0 {
```
Comment (coocood) [inline util/kvencoder/kv_encoder.go]: how about CAS(0, 1)?

Comment (coocood) [reply util/kvencoder/kv_encoder.go]: Or just hold a lock during initGlobal?

File: util/kvencoder/kv_encoder.go [hunk 2]
```diff
@@ -254,7 +240,28 @@ func (e *kvEncoder) initial(dbName string, idAlloc autoid.Allocator) (err error)
...
+
+	if storeGlobal != nil {
+		if err1 := storeGlobal.Close(); err1 != nil {
+			log.Error(errors.ErrorStack(err1))
+		}
+	}
+	if domGlobal != nil {
+		domGlobal.Close()
+	}
 	return nil
```
Comment (holys) [inline util/kvencoder/kv_encoder.go]: return error instead of nil?

**Comment by tiancaiamao** [comment]
PTAL @coocood @holys

File: util/kvencoder/kv_encoder.go [hunk 3]
```diff
@@ -77,27 +78,48 @@ type KvEncoder interface {
 	Close() error
 }
 
+var (
+	// refCount is used to ensure that there is only one domain.Domain instance.
+	refCount    int64
```
Comment (zz-jason) [inline util/kvencoder/kv_encoder.go]: can we remove `refCount`? seems we can use `domGlobal != nil` to replace?

**Comment by zz-jason** [comment]
@tiancaiamao Please add some proper labels for this PR.

**Comment by coocood** [comment]
LGTM

File: util/kvencoder/kv_encoder.go (hunk 3 above)
Comment (tiancaiamao) [reply util/kvencoder/kv_encoder.go]: No, we can't

Comment (tiancaiamao) [reply util/kvencoder/kv_encoder.go]: When refCount decrease to 0, we set `domGlobal = nil`

Comment (zz-jason) [reply util/kvencoder/kv_encoder.go]: So `refCount == 0` is equal to `domGolbal == nil`?

Comment (tiancaiamao) [reply util/kvencoder/kv_encoder.go]: Nope

**Comment by holys** [comment]
LGTM

**Pull request merged**